dbms/src/DataStreams/ParallelParsingBlockInputStream.h

```diff
@@ -0,0 +1,229 @@
+        {
+            buffers.emplace_back(std::make_unique<ReadBuffer>(segments[i].memory.data(), segments[i].used_size, 0));
+            working_field.emplace_back(builder.input_processor_creator, builder.input_creator_params, segments[i], *buffers[i], blocks[i], exceptions[i]);
+        }
+
+        segmentator_thread = ThreadFromGlobalPool([this] { segmentatorThreadFunction(); });
+    }
+
+    String getName() const override { return "ParallelParsing"; }
+
+    ~ParallelParsingBlockInputStream() override
+    {
+        {
+            std::unique_lock lock(mutex);
+            segmentator_condvar.notify_all();
+            reader_condvar.notify_all();
+        }
+
+        if (segmentator_thread.joinable())
+            segmentator_thread.join();
+
+        try
+        {
+            pool.wait();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+    }
+
+    void cancel(bool kill) override
+    {
+        if (kill)
+            is_killed = true;
+        bool old_val = false;
+        if (!is_cancelled.compare_exchange_strong(old_val, true))
+            return;
+
+        for (auto& unit: working_field)
+            unit.reader->cancel(kill);
+
+        {
+            std::unique_lock lock(mutex);
+            segmentator_condvar.notify_all();
+            reader_condvar.notify_all();
+        }
+
+        segmentator_thread.join();
+
+        try
+        {
+            pool.wait();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+
+    }
+
+    Block getHeader() const override
+    {
+        return working_field.at(0).reader->getHeader();
+    }
+
+protected:
+    void readPrefix() override {}
+
+    //Reader routine
+    Block readImpl() override;
+
+    const BlockMissingValues & getMissingValues() const override
+    {
+        std::lock_guard missing_values_lock(missing_values_mutex);
+        return block_missing_values;
+    }
+
+private:
+
+    const std::atomic<size_t> max_threads_to_use;
+    const size_t min_chunk_size;
+
+    std::atomic<bool> is_exception_occured{false};
+
+    BlockMissingValues block_missing_values;
+    mutable std::mutex missing_values_mutex;
+
+    // Original ReadBuffer to read from.
+    ReadBuffer & original_buffer;
+
+    //Non-atomic because it is used in one thread.
+    size_t reader_ticket_number{0};
+    size_t segmentator_ticket_number{0};
+
+    std::mutex mutex;
+    std::condition_variable reader_condvar;
+    std::condition_variable segmentator_condvar;
+
+    // There are multiple "parsers", that's why we use thread pool.
+    ThreadPool pool;
+    // Reading and segmentating the file
+    ThreadFromGlobalPool segmentator_thread;
+
+    // Function to segment the file. Then "parsers" will parse that segments.
+    FormatFactory::FileSegmentationEngine file_segmentation_engine;
+
+    enum ProcessingUnitStatus
+    {
+        READY_TO_INSERT,
```

Comment:
Is this actually an empty unit (i.e. has no file segment and no parsed block, and can be reused)?

Reply:
It means exactly that unit is ready to insert. It may contains previous junk file segment.